File: tests/python/calculator.test.py
"""
Calculator module tests
"""

import sys
import os

# Add the src directory to the path so we can import the calculator module
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../src/python'))

from calculator import sum, subtract, multiply, divide

def test_sum():
    """Test the sum function"""
    print("Testing sum function...")
    
    # Basic tests
    assert sum(2, 3) == 5, "2 + 3 should equal 5"
    print("✓ sum(2, 3) == 5")
    
    assert sum(-1, 1) == 0, "-1 + 1 should equal 0"
    print("✓ sum(-1, 1) == 0")
    
    assert sum(0, 0) == 0, "0 + 0 should equal 0"
    print("✓ sum(0, 0) == 0")
    
    # Test error handling
    try:
        sum('a', 5)
        assert False, "sum should raise TypeError for non-number input"
    except TypeError:
        print("✓ sum raises TypeError for non-number input")

def test_subtract():
    """Test the subtract function"""
    print("\nTesting subtract function...")
    
    assert subtract(5, 3) == 2, "5 - 3 should equal 2"
    print("✓ subtract(5, 3) == 2")
    
    assert subtract(0, 5) == -5, "0 - 5 should equal -5"
    print("✓ subtract(0, 5) == -5")

def test_multiply():
    """Test the multiply function"""
    print("\nTesting multiply function...")
    
    assert multiply(3, 4) == 12, "3 * 4 should equal 12"
    print("✓ multiply(3, 4) == 12")
    
    assert multiply(-2, 3) == -6, "-2 * 3 should equal -6"
    print("✓ multiply(-2, 3) == -6")

def test_divide():
    """Test the divide function"""
    print("\nTesting divide function...")
    
    assert divide(10, 2) == 5, "10 / 2 should equal 5"
    print("✓ divide(10, 2) == 5")
    
    assert divide(7, 2) == 3.5, "7 / 2 should equal 3.5"
    print("✓ divide(7, 2) == 3.5")
    
    # Test division by zero
    try:
        divide(5, 0)
        assert False, "divide should raise ZeroDivisionError for division by zero"
    except ZeroDivisionError:
        print("✓ divide raises ZeroDivisionError for division by zero")

if __name__ == "__main__":
    try:
        test_sum()
        test_subtract()
        test_multiply()
        test_divide()
        print("\nAll tests passed! ✓")
    except Exception as e:
        print(f"Test failed with error: {e}")
        sys.exit(1)

File: tests/js/calculator.test.js
/**
 * Unit tests for the calculator module
 * Uses a more modern test structure while maintaining the custom framework for now
 */

import { sum, MathOperations } from '../../src/js/calculator.js';

// Simple test framework implementation
function test(description, fn) {
    try {
        fn();
        console.log(`✓ ${description}`);
        return true;
    } catch (error) {
        console.error(`✗ ${description}: ${error.message}`);
        return false;
    }
}

// Test results tracker
let passedTests = 0;
let totalTests = 0;

function assertEqual(actual, expected, message) {
    totalTests++;
    if (actual !== expected) {
        throw new Error(message || `Expected ${expected}, got ${actual}`);
    }
    passedTests++;
}

function assertThrows(fn, expectedError, message) {
    totalTests++;
    try {
        fn();
        throw new Error(message || `Expected function to throw an error`);
    } catch (error) {
        if (error.constructor.name !== expectedError) {
            throw new Error(message || `Expected error ${expectedError}, got ${error.constructor.name}`);
        }
        passedTests++;
    }
}

// Run tests
console.log('Running calculator tests...\n');

// Test the sum function (backward compatibility)
test('sum should add positive numbers correctly', () => {
    assertEqual(sum(2, 3), 5, '2 + 3 should equal 5');
});

test('sum should handle negative numbers', () => {
    assertEqual(sum(-1, 1), 0, '-1 + 1 should equal 0');
});

test('sum should handle decimal numbers', () => {
    const result = sum(2.5, 3.7);
    // For floating point comparison, we'll check if it's close enough
    if (Math.abs(result - 6.2) > 0.0001) {
        throw new Error(`2.5 + 3.7 should approximately equal 6.2, got ${result}`);
    }
    passedTests++;
    totalTests++;
});

test('sum should handle zero', () => {
    assertEqual(sum(0, 0), 0, '0 + 0 should equal 0');
    assertEqual(sum(5, 0), 5, '5 + 0 should equal 5');
    assertEqual(sum(0, 5), 5, '0 + 5 should equal 5');
});

test('sum should throw error for invalid inputs', () => {
    assertThrows(() => sum('a', 5), 'Error', 'Should throw Error for string input');
    assertThrows(() => sum(5, null), 'Error', 'Should throw Error for null input');
});

// Test the MathOperations class methods
test('MathOperations.add should add positive numbers correctly', () => {
    assertEqual(MathOperations.add(2, 3), 5, '2 + 3 should equal 5');
});

test('MathOperations.subtract should subtract numbers correctly', () => {
    assertEqual(MathOperations.subtract(5, 3), 2, '5 - 3 should equal 2');
});

test('MathOperations.multiply should multiply numbers correctly', () => {
    assertEqual(MathOperations.multiply(3, 4), 12, '3 * 4 should equal 12');
});

test('MathOperations.divide should divide numbers correctly', () => {
    assertEqual(MathOperations.divide(10, 2), 5, '10 / 2 should equal 5');
});

test('MathOperations.divide should throw error when dividing by zero', () => {
    assertThrows(() => MathOperations.divide(10, 0), 'Error', 'Should throw Error when dividing by zero');
});

// Print results
console.log(`\nResults: ${passedTests}/${totalTests} tests passed`);

if (passedTests !== totalTests) {
    process.exit(1);
} else {
    console.log('All tests passed!');
}